// @livekit/rtc-node/dist/participant.js
// SPDX-FileCopyrightText: 2024 LiveKit, Inc.
//
// SPDX-License-Identifier: Apache-2.0
import { FfiClient, FfiHandle } from './ffi_client.js';
import { DataPacketKind, PublishDataRequest, PublishTrackRequest, UnpublishTrackRequest, UpdateLocalMetadataRequest, UpdateLocalNameRequest, } from './proto/room_pb.js';
import { LocalTrackPublication, } from './track_publication.js';
export class Participant {
    constructor(owned_info) {
        this.trackPublications = new Map();
        this.info = owned_info.info;
        this.ffi_handle = new FfiHandle(owned_info.handle.id);
    }
    get sid() {
        return this.info.sid;
    }
    get name() {
        return this.info.name;
    }
    get identity() {
        return this.info.identity;
    }
    get metadata() {
        return this.info.metadata;
    }
}
export class LocalParticipant extends Participant {
    constructor() {
        super(...arguments);
        this.trackPublications = new Map();
    }
    async publishData(data, options) {
        let req = new PublishDataRequest({
            localParticipantHandle: this.ffi_handle.handle,
            dataPtr: FfiClient.instance.retrievePtr(data),
            dataLen: BigInt(data.byteLength),
            kind: options.reliable ? DataPacketKind.KIND_RELIABLE : DataPacketKind.KIND_LOSSY,
            topic: options.topic,
        });
        if (options.destination) {
            const sids = options.destination.map((sid) => {
                if (typeof sid == 'string')
                    return sid;
                return sid.sid;
            });
            req.destinationSids = sids;
        }
        let res = FfiClient.instance.request({
            message: { case: 'publishData', value: req },
        });
        let cb = await FfiClient.instance.waitFor((ev) => {
            return ev.message.case == 'publishData' && ev.message.value.asyncId == res.asyncId;
        });
        if (cb.error) {
            throw new Error(cb.error);
        }
    }
    async updateMetadata(metadata) {
        let req = new UpdateLocalMetadataRequest({
            localParticipantHandle: this.ffi_handle.handle,
            metadata: metadata,
        });
        let res = FfiClient.instance.request({
            message: { case: 'updateLocalMetadata', value: req },
        });
        await FfiClient.instance.waitFor((ev) => {
            return ev.message.case == 'updateLocalMetadata' && ev.message.value.asyncId == res.asyncId;
        });
    }
    async updateName(name) {
        let req = new UpdateLocalNameRequest({
            localParticipantHandle: this.ffi_handle.handle,
            name: name,
        });
        let res = FfiClient.instance.request({
            message: { case: 'updateLocalName', value: req },
        });
        await FfiClient.instance.waitFor((ev) => {
            return ev.message.case == 'updateLocalName' && ev.message.value.asyncId == res.asyncId;
        });
    }
    async publishTrack(track, options) {
        let req = new PublishTrackRequest({
            localParticipantHandle: this.ffi_handle.handle,
            trackHandle: track.ffi_handle.handle,
            options: options,
        });
        let res = FfiClient.instance.request({
            message: { case: 'publishTrack', value: req },
        });
        let cb = await FfiClient.instance.waitFor((ev) => {
            return ev.message.case == 'publishTrack' && ev.message.value.asyncId == res.asyncId;
        });
        let track_publication = new LocalTrackPublication(cb.publication);
        track_publication.track = track;
        this.trackPublications.set(track_publication.sid, track_publication);
        return track_publication;
    }
    async unpublishTrack(trackSid) {
        let req = new UnpublishTrackRequest({
            localParticipantHandle: this.ffi_handle.handle,
            trackSid: trackSid,
        });
        let res = FfiClient.instance.request({
            message: { case: 'unpublishTrack', value: req },
        });
        await FfiClient.instance.waitFor((ev) => {
            return ev.message.case == 'unpublishTrack' && ev.message.value.asyncId == res.asyncId;
        });
        let pub = this.trackPublications.get(trackSid);
        pub.track = undefined;
        this.trackPublications.delete(trackSid);
    }
}
export class RemoteParticipant extends Participant {
    constructor(owned_info) {
        super(owned_info);
        this.trackPublications = new Map();
    }
}
